test/test_testing.py

```diff
@@ -2965,40 +3002,55 @@ def test_add_network_endpoint_and_relation_id_do_not_correspond(self):
 
 
 class DBRelationChangedHelper(ops.Object):
-    def __init__(self, parent, key):
+    def __init__(self, parent: ops.Object, key: str):
         super().__init__(parent, key)
-        self.changes = []
+        self.changes: typing.List[typing.Tuple[int, str]] = []
         parent.framework.observe(parent.on.db_relation_changed, self.on_relation_changed)
 
-    def on_relation_changed(self, event):
+    def on_relation_changed(self, event: ops.RelationEvent):
         if event.unit is not None:
             self.changes.append((event.relation.id, event.unit.name))
         else:
-            self.changes.append((event.relation.id, event.app.name))
+            app = event.app
+            assert app is not None
+            self.changes.append((event.relation.id, app.name))
 
 
 class RelationChangedViewer(ops.Object):
     """Track relation_changed events and saves the data seen in the relation bucket."""
 
-    def __init__(self, charm, relation_name):
+    def __init__(self, charm: ops.CharmBase, relation_name: str):
         super().__init__(charm, relation_name)
-        self.changes = []
+        self.changes: typing.List[typing.Dict[str, str]] = []
         charm.framework.observe(charm.on[relation_name].relation_changed, self.on_relation_changed)
 
-    def on_relation_changed(self, event):
+    def on_relation_changed(self, event: ops.RelationEvent):
         if event.unit is not None:
             data = event.relation.data[event.unit]
         else:
-            data = event.relation.data[event.app]
+            app = event.app
+            assert app is not None
+            data = event.relation.data[app]
         self.changes.append(dict(data))
 
 
+if typing.TYPE_CHECKING:
```

Comment:
Any reason these need to be in `if typing.TYPE_CHECKING` blocks? I think we should avoid it when possible for simplicity and consistency, but I'm open to being convinced otherwise.

Reply:
No, I was just trying to match the style elsewhere in the code. Happy to drop them.